### DISCUSSION TITLE
Using the results of a Dataview search to populate `suggester` values

### DISCUSSION THREAD

**Discussion by buttonpushertv** [chhoumann/quickadd#604] [Q&A]
I'm trying to make sense of the code example given for the `suggester` function in a user script:

```
const pickedFile = await params.quickAddApi.suggester(
    (file) => file.basename,
    params.app.vault.getMarkdownFiles()
);
```

Clearly I don't have a full grasp on how JS mapping works. Where is that '(file)' value coming from? Is it the result of the 'getMarkdownFiles()' function?

I am, however, trying to use a Dataview result (pages with a specific tag) as the values to offer in the `suggester`. I'm trying to make use of a set of pages for an RPG campaign. Each of the pages related to a city or town have the tag: ttrpg-burg. The script below is for creating a new item that will be located in a specific Burg. I want to provide the suggester listing if the item is created from a page that isn't already a Burg (if it is created from a Burg page, it just uses that page's `Burg` value.)

Here is what I have (I've added comments to discuss success and failures):

```
module.exports = async (params) => {
    const {quickAddApi: {inputPrompt},} = params;
    const dataviewAPI = params.app.plugins.plugins.dataview.api; \\ activates DV API
    const burgPages = await dataviewAPI.pages("#ttrpg-burg"); \\ creates an array of all those pages
    const burgListing = burgPages.map(obj => obj.Burg); \\ creates a 2nd array of just the Burg names only
    const newItemName = await inputPrompt("Name for the new Item?"); \\ ask for the new Item's name
    propList = params.app.workspace.activeEditor.metadataEditor.properties[2]; \\ capturing this specific property
    if (propList.key !== 'Burg') { \\ if it isn't a Burg, then open the `suggester` dialog
        const newItemLocation = await params.quickAddApi.suggester(
        burgListing, burgListing \\ this *almost* works BUT I know I'm missing something (see below)
        );
    } else { \\ this works to set newItemLocation if its created from an existing Burg page
        newItemLocation = propList.value; 
    };
    params.variables["newItemName"] = newItemName; \\ passing these items back for the template to use
    params.variables["newItemLocation"] = newItemLocation;
};
```
It is very close to working. I just need to figure out how the `suggester` should be fed info to use & then return usable info from it.
I know I'm missing something in the `suggester` function. I've tried a few variations on the burgListing and burgPages, but this is where I do get a listing in the suggester that is, at least, just the burg names - which, BTW, are also the names of the Markdown files for each Burg.

I think I'm just not mapping the burgListing values to that object properly, or I'm not having them map properly within the suggester. I'm guessing, with the code above, even though I'm seeing the listing in the suggester, I'm returning the entire object as the result. With the way it is above, like I said, I get the listing of all the Burg pages (tagged: `#ttrpg-burg`) but it does not do any kind completion to select a choice (and throws an error even if I try to select an item.)

Can someone help me figure out what it is I'm not getting correct? Thanks.

**Comment by buttonpushertv** [comment]
I have made *a little* progress. I still cannot get the Dataview listing to be passed through the `suggester` function, but I have it nearly working now using the `app.vault.getMarkdownFiles()` method, however...

The script is *almost* working. I can't figure out why `newItemNameLocation` is not being set if the first `if` condition is `true`.

If the second `if` condition evaluates as true, then it works correctly.

I added the line at the top - `let newItemLocation = "";` because the variable, once set, holds it's value through subsequent calls. It was succeeding erroneously because that first `if` doesn't seem to pass anything to `newItemLocation` at all. Setting it to `nul`, at least, forces the error condition, but I cannot figure out why it's not setting the value in the first `if` condition.

(FWIW, I have a ton of `console.log` lines in another version and I see that the first condition definitely gets tested properly and I see that `tempItemLocation` *does* get defined and the `.basename` gets properly assigned to `newItemLocation` within the condition - it just *doesn't* make it down to the `params.variables` assignment line at the bottom if the first `if` condition is `true`.)

Here is the script:

```
module.exports = async (params) => {
    let newItemLocation = "";
    const {quickAddApi: {inputPrompt,suggester},} = params;
    const newItemName = await inputPrompt("Name for the new Item?");
    const propList = params.app.workspace.activeEditor.metadataEditor.properties;
    // All of my 'Burg' notes, have identical metadata in the properties
    // The following 'if' tests if a page is a 'Burg' note or not.
    if (propList[1].value !== 'Burg') {
        let tempItemLocation = await suggester(
            (file) => file.basename, 
            params.app.vault.getMarkdownFiles()
            );
        let newItemLocation = tempItemLocation.basename;
    } else if (propList[1].value == 'Burg') {
        newItemLocation = propList[2].value;
    };
    let params.variables["newItemName"] = newItemName;
    let params.variables["newItemLocation"] = newItemLocation;
};
```

And this is the template called by the QuickAdd Macro:

```
---
NoteIcon: Item
Type: Item
Name: {{VALUE:newItemName}}
Campaign: {{VALUE:thisCampaignName}}
Location: {{VALUE:newItemLocation}}
Template: "[[Template - Item]]"
TemplateVersion: 1.0
---

(human readable page content here)
```

Any thoughts?

**Comment by chhoumann** [comment]
Answering this far too late - sorry. You were very close; there are two plain JavaScript bugs in the second script, plus one Dataview quirk in the first.

**Your original `(file)` question:** `suggester`'s first argument is a function QuickAdd calls once for each element of the second argument to build the label shown in the menu. What the suggester *returns* is the picked element of the second array, not the label. So with `getMarkdownFiles()` you get back a `TFile`, and `.basename` gets you its name - you had that part right.

**Bug 1 - why `newItemLocation` never escapes the first `if`:** `let newItemLocation = tempItemLocation.basename;` inside the block declares a *new* block-scoped variable that shadows the outer one and disappears at the closing brace. Drop the `let` so it assigns the outer variable:

```js
newItemLocation = tempItemLocation.basename;
```

**Bug 2:** `let params.variables["newItemName"] = ...` is invalid JavaScript - `let` declares a new variable, but `params.variables[...]` is a property assignment. Remove `let` from both of those lines.

**The Dataview version:** `dv.pages()` returns a DataArray, not a plain Array (and `.map()` on it returns another DataArray), which is why feeding it to `suggester` misbehaved. Convert with `.array()`:

```js
const burgs = params.app.plugins.plugins.dataview.api
    .pages("#ttrpg-burg")
    .map((p) => p.file.name)
    .array();
const picked = await suggester(burgs, burgs);
```

Passing the same string array as both arguments means the resolved value *is* the picked name - no `.basename` step needed.

Putting it all together:

```js
module.exports = async (params) => {
    const { quickAddApi: { inputPrompt, suggester } } = params;
    const newItemName = await inputPrompt("Name for the new Item?");
    const propList = params.app.workspace.activeEditor.metadataEditor.properties;
    let newItemLocation;
    if (propList[1].value !== "Burg") {
        const burgs = params.app.plugins.plugins.dataview.api
            .pages("#ttrpg-burg")
            .map((p) => p.file.name)
            .array();
        newItemLocation = await suggester(burgs, burgs);
    } else {
        newItemLocation = propList[2].value;
    }
    params.variables["newItemName"] = newItemName;
    params.variables["newItemLocation"] = newItemLocation;
};
```

API reference: https://quickadd.obsidian.guide/docs/QuickAddAPI
